Various optimizations to protocols to improve performance

gabearro opened the pull request:
Replace `Mutex<HashMap>` with `DashMap` to improve concurrent session access and simplify synchronization. 
Switch from manual locking to DashMap's atomic operations in `RBC` implementations and restructure session management for efficiency.
Optimize `lagrange_interpolate` with parallel computation for large inputs using crossbeam. 
Simplify and streamline `NetworkErrorCode` handling. 
Enhance triple generation with parallel processing for batch execution. 
Include batched `RanSha` functionality support. Integrate `DashMap` and `crossbeam` as new dependencies.

gabearro committed:
Refactor: Replace `Mutex<HashMap>` with `DashMap` to improve concurrent session access and simplify synchronization. Switch from manual locking to DashMap's atomic operations in `RBC` implementations and restructure session management for efficiency. Optimize `lagrange_interpolate` with parallel computation for large inputs using crossbeam. Simplify and streamline `NetworkErrorCode` handling. Enhance triple generation with parallel processing for batch execution. Include batched `RanSha` functionality support. Integrate `DashMap` and `crossbeam` as new dependencies.

Mikerah commented on mpc/src/common/mod.rs:
```diff
@@ -79,6 +81,52 @@ pub fn lagrange_interpolate<F: FftField>(
     let n = x_vals.len();
+
+    // For small inputs, use sequential computation to avoid thread overhead
+    if n <= 4 {
```
Do we have an empirical way to see what n should actually be here? Like an elbow curve?

Mikerah commented on mpc/src/honeybadger/share_gen/batched_share_gen.rs:
```diff
@@ -0,0 +1,520 @@
+//! - Original protocol: 1 secret per party → 3 output shares per run
+//! - Batched protocol: 512 secrets per party → 1536 output shares per run
+//!
+//! This reduces the number of protocol runs by ~500x for large preprocessing needs.
```
🔥🔥🔥

gabearro committed:
Refactor: Enforce deterministic session ID order for preprocessing outputs to prevent mismatched results across parties due to network delays. Transition to using `HashMap` for collected outputs and ensure sorted processing. Add diagnostic logging for interpolation steps and optimize robust interpolation error handling.

Mikerah commented on mpc/src/honeybadger/mod.rs:
```diff
@@ -1095,24 +1095,22 @@ where
+                // With network delays, different parties may receive outputs in different orders.
+                // If we add in arrival order, parties get mismatched triples causing interpolation failures.
                 let triple_channel = self.outputchannels.triple_channel.clone();
+                let mut collected_triples: HashMap<SessionId, Vec<ShamirBeaverTriple<F>>> = HashMap::new();
```
HashMaps in Rust don't enforce an ordering. Shouldn't another data structure be used?

Mikerah commented on mpc/src/honeybadger/mod.rs:
```diff
@@ -1355,22 +1375,22 @@ where
+        // IMPORTANT: We must add outputs in deterministic session ID order, not arrival order.
+        // With network delays, different parties may receive outputs in different orders.
+        // If we add in arrival order, parties get mismatched shares causing interpolation failures.
+        let mut collected_outputs: HashMap<SessionId, Vec<RobustShare<F>>> = HashMap::new();
```
Same here. HashMap doesn't enforce an ordering.

gabearro committed:
Add benchmarking for Lagrange interpolation and expose internal methods for testing. Update threshold for sequential computation and integrate Criterion for performance analysis.

gabearro committed:
Drop unused `store` to ensure proper resource cleanup in batched share generation.

cadaniluk commented on mpc/src/honeybadger/mod.rs:
```diff
@@ -1020,62 +1054,89 @@ where
+
+                while collected < num_triple_batches {
+                    if let Some(sid) = triple_channel.lock().await.recv().await {
+                        if triple_session_ids.contains(&sid) && !collected_triples.contains_key(&sid) {
```
Do we really need these checks here? If we get duplicate session IDs in the output channel or a non-existing session ID, then the subprotocol is broken or there are too many malicious nodes.

gabearro committed:
Integrate batched RanDouSha protocol for large-scale double share generation, including support for batched message processing, deterministic session ID generation, and enhanced diagnostic logging. Refactor existing RanDouSha logic to accommodate both regular and batched modes, ensuring efficient share collection and output handling.

gabearro committed:
Add large-scale performance and preprocessing stress tests for MPC protocols

- Introduced extensive benchmarking for core operations, including share generation, recovery, and share multiplication.
- Added tests for Lagrange interpolation, robust vs. non-robust recovery, and FFT vs. Lagrange for optimization analysis.
- Developed stress tests for RanDouSha, RanSha, and batched RanSha protocols to evaluate scalability and throughput under heavy workloads.
- Enhanced diagnostics with detailed performance summaries, memory usage estimation, and bottleneck analysis.

gabearro committed:
Add batch field and polynomial operations for optimized MPC computations

- Introduced `batch_ops` module with efficient implementations of common field operations (batched multiplication, addition, subtraction, scalar multiplication), as well as polynomial operations such as coefficient-wise addition, scalar multiplication, and summation.
- Optimized Vandermonde matrix computation and evaluation functions (single-point and batched) with parallel processing and Montgomery's trick for batch inversion to reduce field inversion overhead.
- Updated Lagrange interpolation and robust interpolation to leverage batched operations, improving performance by parallelizing computation and minimizing redundant field operations.
- Refactored double-share generation logic in RanDouSha and Batched RanDouSha to detect readiness for output phase earlier, avoid redundant locking, and better handle partially processed states.
- Enhanced logging granularity by integrating `trace` logs for finer debugging.

gabearro committed:
Improve async executor efficiency by introducing periodic yielding in batched share generation and RanDouSha; adjust test parameters for large-scale workloads

Mikerah commented:
Closing because a new PR #71 supercedes this PR
